tests/ranged_balance.cpp

```diff
@@ -0,0 +1,275 @@
+        os << " " << *ii;
+    }
+    os << " ]";
+    return os;
+}
+
+std::ostream &operator<<( std::ostream &stream, const dispersion_sources &sources )
+{
+    if( !sources.normal_sources.empty() ) {
+        stream << "Normal: " << sources.normal_sources << std::endl;
+    }
+    if( !sources.linear_sources.empty() ) {
+        stream << "Linear: " << sources.linear_sources << std::endl;
+    }
+    if( !sources.multipliers.empty() ) {
+        stream << "Mult: " << sources.multipliers << std::endl;
+    }
+    return stream;
+}
+
+static void arm_shooter( npc &shooter, std::string gun_type, std::vector<std::string> mods = {} )
+{
+    shooter.remove_weapon();
+
+    itype_id gun_id( gun_type );
+    // Give shooter a loaded gun of the requested type.
+    item &gun = shooter.i_add( item( gun_id ) );
+    const itype_id ammo_id = gun.ammo_default();
+    if( gun.magazine_integral() ) {
+        item &ammo = shooter.i_add( item( ammo_id, calendar::turn, gun.ammo_capacity() ) );
+        REQUIRE( gun.is_reloadable_with( ammo_id ) );
+        REQUIRE( shooter.can_reload( gun, ammo_id ) );
+        gun.reload( shooter, item_location( shooter, &ammo ), gun.ammo_capacity() );
+    } else {
+        const itype_id magazine_id = gun.magazine_default();
+        item &magazine = shooter.i_add( item( magazine_id ) );
+        item &ammo = shooter.i_add( item( ammo_id, calendar::turn, magazine.ammo_capacity() ) );
+        REQUIRE( magazine.is_reloadable_with( ammo_id ) );
+        REQUIRE( shooter.can_reload( magazine, ammo_id ) );
+        magazine.reload( shooter, item_location( shooter, &ammo ), magazine.ammo_capacity() );
+        gun.reload( shooter, item_location( shooter, &magazine ), magazine.ammo_capacity() );
+    }
+    for( auto mod : mods ) {
+        gun.contents.push_back( item( itype_id( mod ) ) );
+    }
+    shooter.wield( gun );
+}
+
+static void equip_shooter( npc &shooter, std::vector<std::string> apparel )
+{
+    tripoint shooter_pos( 60, 60, 0 );
+    shooter.setpos( shooter_pos );
+    shooter.worn.clear();
+    shooter.inv.clear();
+    for( const std::string article : apparel ) {
+        shooter.wear_item( item( article ) );
+    }
+}
+
+std::array<double, 5> accuracy_levels = {{ accuracy_grazing, accuracy_standard, accuracy_goodhit, accuracy_critical, accuracy_headshot }};
+
+static std::array<statistics, 5> firing_test( dispersion_sources dispersion, int range,
+        std::array<double, 5> thresholds )
+{
+    std::array<statistics, 5> firing_stats;
+    bool threshold_within_confidence_interval = false;
+    do {
+        projectile_attack_aim aim = projectile_attack_roll( dispersion, range, 0.5 );
+        threshold_within_confidence_interval = false;
+        for( int i = 0; i < accuracy_levels.size(); ++i ) {
+            firing_stats[i].add( aim.missed_by < accuracy_levels[i] );
+            if( thresholds[i] == -1 ) {
+                continue;
+            }
+            if( firing_stats[i].n() < 100 ) {
```

Comment:
I'm not sure if I understand what is happening here.
If the test has been performed for less than 100 samples, it should end?

Comments would really help a lot here.